mysite/psa/pipeline.py

```diff
@@ -0,0 +1,296 @@
+                        )
+                    else:
+                        new = users[0]
+                if not new:
+                    user.username = details.get('username')
+                    user.first_name = ''
+                    user.save()
+                else:
+                    tmp_user = user
+                    logout(strategy.request)
+                    user = new
+                    user.backend = 'django.contrib.auth.backends.ModelBackend'
+                    login(strategy.request, user)
+                    union_merge(tmp_user, user)
+                    tmp_user.delete()
+                    return {'user': user}
+            except IntegrityError as e:
+                _id = int(time.mktime(datetime.now().timetuple()))
+                user.username = details.get('username') + str(_id)
+                user.save()
+    elif user and social and social.user != user:
+        confirm = strategy.request.POST.get('confirm')
+        if confirm and confirm == 'no':
+            raise AuthException(
+                backend,
+                'You interrupted merge process.'
+            )
+        elif (not user.get_full_name() == social.user.get_full_name() and
+              not strategy.request.POST.get('confirm') and
+              not user.email == social.user.email):
+            return render_to_response('ct/person.html', {
+                'available_backends': load_backends(settings.AUTHENTICATION_BACKENDS),
+                'request': strategy.request,
+                'next': strategy.request.POST.get('next') or '',
+                'target_name': social.user.get_full_name(),
+                'own_name': user.get_full_name(),
+                'person': user,
+                'merge_confirm': True
+            }, RequestContext(strategy.request))
+        elif (user.get_full_name() == social.user.get_full_name()
+              or confirm and confirm == 'yes'
+              or user.email == social.user.email):
+            union_merge(social.user, user)
+            social_merge(social.user, user)
+            social.user.delete()
+            social.user = user
+            return {'user': user,
+                    'social': social}
+
+
+def not_allowed_to_merge(user, social_user):
+    """Check if two users are allowed to merge
+
+    Check all social-auth from two users to predict providers intersection.
+    """
+    user_auths = set([i.provider for i in user.social_auth.all()])
+    social_user_auths = set([i.provider for i in social_user.social_auth.all()])
+
+    return user_auths.intersection(social_user_auths)
+
+
+def social_user(backend, uid, user=None, *args, **kwargs):
+    """Search for UserSocialAuth"""
+    provider = backend.name
+    social = backend.strategy.storage.user.get_social_auth(provider, uid)
+    if social:
+        if user and social.user != user and not user.groups.filter(name='Temporary').exists():
```

Comment:
Remainder: let's discuss this. (create UserState class and define states, if possible)